src/simulation/m_hyperelastic.fpp

```diff
@@ -0,0 +1,336 @@
+
+    use m_derived_types        !< Definitions of the derived types
+
+    use m_global_parameters    !< Definitions of the global parameters
+
+    use m_variables_conversion !< State variables type conversion procedures
+
+    use m_finite_differences
+
+    ! ==========================================================================
+
+    implicit none
+
+    private; public :: s_hyperelastic_rmt_stress_update, &
+ s_initialize_hyperelastic_module, &
+ s_finalize_hyperelastic_module
+
+    !> @name Abstract interface for creating function pointers
+    !> @{
+    abstract interface
+
+        !> @name Abstract subroutine for the infinite relaxation solver
+        !> @{
+        subroutine s_abstract_hyperelastic_solver(btensor, q_prim_vf, G, j, k, l)
+            !!!!$acc routine seq
+            import :: scalar_field, sys_size, b_size
+            type(scalar_field), dimension(sys_size), intent(inout) :: q_prim_vf
+            type(scalar_field), dimension(b_size), intent(inout) :: btensor
+            real(kind(0d0)), intent(in) :: G
+            integer, intent(in) :: j, k, l
+
+        end subroutine s_abstract_hyperelastic_solver
+        !> @}
+
+    end interface
+    !> @}
+
+    !! The btensor at the cell-interior Gaussian quadrature points.
+    !! These tensor is needed to be calculated once and make the code DRY.
+    type(vector_field) :: btensor !<
+!$acc declare create(btensor)
+
+    real(kind(0d0)), allocatable, dimension(:, :) :: fd_coeff_x
+    real(kind(0d0)), allocatable, dimension(:, :) :: fd_coeff_y
+    real(kind(0d0)), allocatable, dimension(:, :) :: fd_coeff_z
+    !$acc declare create(fd_coeff_x,fd_coeff_y,fd_coeff_z)
+    real(kind(0d0)), allocatable, dimension(:) :: Gs
+    !$acc declare create(Gs)
+
+contains
+
+    !>  The following subroutine handles the calculation of the btensor.
+        !!   The calculation of the btensor takes qprimvf.
+        !! @param q_prim_vf Primitive variables
+        !! @param btensor is the output
+        !! calculate the grad_xi, grad_xi is a nxn tensor
+        !! calculate the inverse of grad_xi to obtain F, F is a nxn tensor
+        !! calculate the FFtranspose to obtain the btensor, btensor is nxn tensor
+        !! btensor is symmetric, save the data space
+    subroutine s_initialize_hyperelastic_module
+        integer :: i !< generic iterator
+
+        @:ALLOCATE(btensor%vf(1:b_size))
+        do i = 1, b_size
+            @:ALLOCATE(btensor%vf(i)%sf(0:m, 0:n, 0:p))
+        end do
+        @:ACC_SETUP_VFs(btensor)
+
+        @:ALLOCATE(Gs(1:num_fluids))
+        !$acc loop seq
+        do i = 1, num_fluids
+            Gs(i) = fluid_pp(i)%G
+        end do
+        !$acc update device(Gs)
+
+        @:ALLOCATE_GLOBAL(fd_coeff_x(-fd_number:fd_number, 0:m))
+        if (n > 0) then
+            @:ALLOCATE_GLOBAL(fd_coeff_y(-fd_number:fd_number, 0:n))
+        end if
+        if (p > 0) then
+            @:ALLOCATE_GLOBAL(fd_coeff_z(-fd_number:fd_number, 0:p))
```

Comment:
remove all allocate_global and replace with `@:ALLOCATE`, alocate_global no longer exists